tests/integration/devfile/cmd_devfile_push_test.go

```diff
@@ -32,254 +32,198 @@ var _ = Describe("odo devfile push command tests", func() {
+			BeforeEach(func() {
 				helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-variables.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
-
-				output := helper.Cmd("odo", "push").ShouldPass().Out()
-				Expect(output).To(ContainSubstring("Changes successfully pushed to component"))
-
-				envVars := commonVar.CliRunner.GetEnvsDevFileDeployment(cmpName, "app", commonVar.Project)
-				// check if the env variable has a correct value. This value was substituted from in devfile from variable
-				Expect(envVars["FOO"]).To(Equal("bar"))
-
 			})
-		})
-	})
-
-	Context("verify files are correctly synced", func() {
-
-		It("should not build when no changes are detected in the directory and build when a file change is detected", func() {
-			utils.ExecPushToTestFileChanges(commonVar.Context, cmpName, commonVar.Project)
-		})
 
-		When("local files are created and deleted ", func() {
-			It("should correctly propagate changes to the container", func() {
-				newFilePath := filepath.Join(commonVar.Context, "foobar.txt")
-				newDirPath := filepath.Join(commonVar.Context, "testdir")
-				utils.ExecPushWithNewFileAndDir(commonVar.Context, cmpName, commonVar.Project, newFilePath, newDirPath)
+			When("doing odo push", func() {
 
-				// Check to see if it's been pushed (foobar.txt abd directory testdir)
-				podName := commonVar.CliRunner.GetRunningPodNameByComponent(cmpName, commonVar.Project)
+				BeforeEach(func() {
+					output = helper.Cmd("odo", "push").ShouldPass().Out()
+				})
 
-				stdOut := commonVar.CliRunner.ExecListDir(podName, commonVar.Project, sourcePath)
-				helper.MatchAllInOutput(stdOut, []string{"foobar.txt", "testdir"})
+				It("should pass", func() {
 
-				// Now we delete the file and dir and push
-				helper.DeleteDir(newFilePath)
-				helper.DeleteDir(newDirPath)
-				helper.Cmd("odo", "push", "--project", commonVar.Project, "-v4").ShouldPass()
+					Expect(output).To(ContainSubstring("Changes successfully pushed to component"))
+				})
+				It("should check if the env variable has a correct value", func() {
 
-				// Then check to see if it's truly been deleted
-				stdOut = commonVar.CliRunner.ExecListDir(podName, commonVar.Project, sourcePath)
-				helper.DontMatchAllInOutput(stdOut, []string{"foobar.txt", "testdir"})
+					envVars := commonVar.CliRunner.GetEnvsDevFileDeployment(cmpName, "app", commonVar.Project)
+					// check if the env variable has a correct value. This value was substituted from in devfile from variable
+					Expect(envVars["FOO"]).To(Equal("bar"))
+				})
 			})
 		})
+		When("doing odo push", func() {
+			BeforeEach(func() {
+				helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
+				helper.Cmd("odo", "push", "--project", commonVar.Project).ShouldPass()
+			})
+			When("doing odo push again", func() {
```

Comment:
IMO we don't need `when` for https://github.com/openshift/odo/pull/4924/files#diff-88d48f926f1685532ec16b131967675b6c60db3c311fafcec09b0339102d8d26R154-R177 because each when uses only one spec. Adding just `It` would be enough. Correct me if I am wrong

Reply:
So, as per discussion with @feloy, we will use `when` to describe each step that we take, and `BeforeEach` and `AfterEach` will we used to implement the `when` and lastly `It` spec will be used for mainly for verifying.

Reply:
> So, as per discussion with @feloy, we will use `when` to describe each step that we take, and `BeforeEach` and `AfterEach` will we used to implement the `when` and lastly `It` spec will be used for mainly for verifying.

Agreed, but IIRC we are also trying to reduce nested cases as much as possible. IMO such cases are the one where we can think of reducing it.

Reply:
@feloy WDYT?

Reply:
I would prefer to keep the `when` here, so the different steps are very clear.  There are not too many nested `when`s. Also, it gives the possibility to add more `It`s in this `When` later easily.

Reply:
> I would prefer to keep the when here, so the different steps are very clear. There are not too many nested whens. Also, it gives the possibility to add more Its in this When later easily.

I agree with the point here that it doesn't have too many nested whens. hence we can consider this scenario to be implemented with when. However if nested level goes deeper like 4 or 5 layer of nesting, this will lead us into a problem. But yes in this case I don't have strong disagreement.